exonum/src/storage/proof_map_index/proof.rs

```diff
@@ -385,93 +381,85 @@ impl<V: fmt::Debug> BranchProofNode<V> {
 
     fn validate_consistency<'a>(
         &'a self,
-        parent_slice: &DBKey,
-        searched_slice: &DBKey,
+        parent_path: &ProofPath,
+        searched_path: &ProofPath,
     ) -> Result<Option<&'a V>, Error> {
         use self::BranchProofNode::*;
 
         // if we inspect sub-proofs of a proof
         let res: Option<&V> = match *self {
             LeftBranch {
                 left_node: ref proof,
-                left_key: ref left_slice_key,
-                right_key: ref right_slice_key,
+                left_key: ref left_path,
+                right_key: ref right_path,
                 ..
             } => {
-                let mut left_slice = left_slice_key.clone();
-                left_slice.set_from(0);
-                let mut right_slice = right_slice_key.clone();
-                right_slice.set_from(0);
-                if !left_slice.starts_with(parent_slice) || !right_slice.starts_with(parent_slice) {
+                let mut left_path = left_path.start_from(0);
+                let mut right_path = right_path.start_from(0);
+                if !left_path.starts_with(parent_path) || !right_path.starts_with(parent_path) {
                     return Err(Error::new(format!(
                         "Proof is inconsistent with itself: Proof: \
-                                                   {:?} . Parent slice: {:?}",
+                         {:?} . Parent path: {:?}",
                         self,
-                        parent_slice
+                        parent_path
                     )));
                 }
-                if !searched_slice.starts_with(&left_slice) {
+                if !searched_path.starts_with(&left_path) {
                     return Err(Error::new(format!(
                         "Proof is inconsistent with searched_key: \
-                                                   {:?}. Proof: {:?}",
-                        searched_slice,
+                         {:?}. Proof: {:?}",
+                        searched_path,
                         self
                     )));
                 }
-                proof.validate_consistency(&left_slice, searched_slice)?
+                proof.validate_consistency(&left_path, searched_path)?
             }
             RightBranch {
                 right_node: ref proof,
-                left_key: ref left_slice_key,
-                right_key: ref right_slice_key,
+                left_key: ref left_path,
+                right_key: ref right_path,
                 ..
             } => {
-                let mut left_slice = left_slice_key.clone();
-                left_slice.set_from(0);
-                let mut right_slice = right_slice_key.clone();
-                right_slice.set_from(0);
-                if !left_slice.starts_with(parent_slice) || !right_slice.starts_with(parent_slice) {
+                let mut left_path = left_path.start_from(0);
```

Comment:
Nitpick: Would it make sense to implement `Clone` for `ProofPath` instead of using `.start_from(0)`? This seems a bit hacky to me.

Reply:
In my opinion clone that produce not an exact copy looks confusing.

Reply:
Got it; for some reason I thought that the path here necessarily starts from 0. Although this begs another question - what is the expected behavior when `start_from` shifts the starting position to the left - it could be dealt with later, as discussed privately.